### DISCUSSION TITLE
How to mutate the Snapshot State of a ProjectResource?

### DISCUSSION THREAD

**Discussion by wertzui** [microsoft/aspire#4964] [Q&A]
I'm trying to add my own `IDistributedApplicationLifecycleHook` to periodically check the health endpoints of resources and change the `CustomResourceSnapshot.State` according to the outcome.

However for Project Resources, the displayed state on the dashboard seems to not be the one from the original `IResource`, but rather from the one which is then running in the DCP.

I tried listening on `ResourceNotificationService.WatchAsync` to find out when an update for the resource with the original resources name is coming in, but it does not seem to work.

How can I add an `IDistributedApplicationLifecycleHook` to the project resource, and get the DCP-resource from it so I can publish state updates of the DCP resource?

**Comment by davidfowl** [comment]
You can’t really mutate the state of project resources, they’ll be overwritten by dcp updates (multiple sources of truth)

**Reply by wertzui** [reply]
Did you mean to write "can't"?

Which resource is the one displayed on the Dashboard?
The project resource, or the DCP resource?

**Reply by davidfowl** [reply]
Yep can't. 

> Which resource is the one displayed on the Dashboard?

It's a bit complex but just know you can't do it from outside of the codebase reliably. Projects and executable resources are owned by the [`ApplicationExecutor`](https://github.com/dotnet/aspire/blob/main/src/Aspire.Hosting/Dcp/ApplicationExecutor.cs). They are updated via the [`ResourceNotificationService`](https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.applicationmodel.resourcenotificationservice?view=dotnet-aspire-8.0). 

That said, updates are read from DCP and snapshots are pushed via that API. The problem isn't so much the API but the fact that multiple mutators will be trying to update the status and it won't work well as they can't be synchronized.
